app/src/org/commcare/android/nfc/NfcReadActivity.java

```diff
@@ -0,0 +1,119 @@
+package org.commcare.android.nfc;
+
+import android.annotation.TargetApi;
+import android.content.Intent;
+import android.nfc.FormatException;
+import android.nfc.NdefMessage;
+import android.nfc.NdefRecord;
+import android.nfc.Tag;
+import android.nfc.tech.Ndef;
+import android.os.Build;
+import android.os.Bundle;
+import android.util.Pair;
+
+import org.commcare.android.javarosa.IntentCallout;
+
+import java.io.IOException;
+
+/**
+ * When this activity is in the foreground, any NFC tag scanned by the device that is of a format
+ * CommCare recognizes (as determined by the filters in NfcActivity::setReadyToHandleTag) will be
+ * consumed by this activity. If the tag is of the specified/expected type(s), then this activity
+ * will attempt to read the tag. If the read is successful, the value read will be stored as an
+ * extra in the activity's result intent under the key "odk_intent_data", where it can be accessed
+ * by other activities. Otherwise, an appropriate error toast will be shown.
+ *
+ * @author Aliza Stone
+ */
+@TargetApi(Build.VERSION_CODES.JELLY_BEAN)
+public class NfcReadActivity extends NfcActivity {
+
+    private String valueRead;
+    private String[] acceptableTypes;
+
+    @Override
+    protected void initFields() {
+        super.initFields();
+        this.acceptableTypes = parseTypes(getIntent());
+    }
+
+    protected static String[] parseTypes(Intent i) {
+        String typesString = i.getStringExtra(NFC_PAYLOAD_MULT_TYPES_ARG);
+        if (typesString == null || typesString.equals("")) {
+            String singleType = i.getStringExtra(NFC_PAYLOAD_SINGLE_TYPE_ARG);
+            if (singleType == null || singleType.equals("")) {
+                return null;
+            } else {
+                return new String[]{ singleType };
+            }
+        } else {
+            return typesString.split(" ");
+        }
+    }
+
+    @Override
+    protected boolean requiredFieldsMissing() {
+        if (this.acceptableTypes == null || this.acceptableTypes.length == 0) {
+            finishWithErrorToast("nfc.read.no.type");
+            return true;
+        }
+        return false;
+    }
+
+    @Override
+    protected void dispatchActionOnTag(Tag tag) {
+        readFromNfcTag(tag);
+    }
+
+    private void readFromNfcTag(Tag tag) {
+        Ndef ndefObject = Ndef.get(tag);
+        try {
+            ndefObject.connect();
+            NdefMessage msg = ndefObject.getNdefMessage();
+            if (msg == null) {
+                finishWithErrorToast("nfc.read.no.data");
+                return;
+            }
+            NdefRecord firstRecord = msg.getRecords()[0];
+            Pair<String, Boolean> resultAndSuccess =
```

Comment:
Might be worth mentioning in a method comment for `readValueFromRecord` that the first result is either the successfully read value *or* the error message locale key

Reply:
Yes definitely